client/src/routes/login/+page.svelte

```diff
@@ -1,46 +1,48 @@
 <script lang="ts">
-  import { z } from 'zod';
-  import { superForm, setMessage } from "sveltekit-superforms/client";
-    import { goto } from '$app/navigation';
+  import { z } from "zod";
+  import {
+    superForm,
+    superValidateSync,
+    setMessage,
+  } from "sveltekit-superforms/client";
+  import { goto } from "$app/navigation";
+  import type { ErrorResponse } from "bindings/ErrorResponse";
 
-  let form_data;
-
-  const userSchema = z.object({
+  const loginSchema = z.object({
     email: z.string().email(),
     password: z.string().min(8),
   });
 
-  const { form, errors, message, constraints, enhance } = superForm(
-    form_data,
+  const { form, message, errors, constraints, enhance } = superForm(
+    superValidateSync(loginSchema),
     {
       SPA: true,
-      validators: userSchema,
+      validators: loginSchema,
       onUpdate: async ({ form }) => {
-        console.log(form.valid);
         if (form.valid) {
-          setMessage(form, "Valid data!");
           const res = await fetch("http://localhost:6060/api/v1/users/login", {
             credentials: "include",
             method: "POST",
             headers: {
               "Content-Type": "application/json",
             },
-            body: JSON.stringify(form.data)
+            body: JSON.stringify(form.data),
           });
-          await goto("/");
+          if (res.status >= 400) {
+            const err: ErrorResponse = await res.json();
+            setMessage(form, err.error);
```

Comment:
can this be done like the registeration form?

```rust
            const err: ErrorResponse = await res.json();
            const errField = err.error.includes("email") ? "email" : "password"; 
            setError(form, errField, err.error);
```

Reply:
backend returns `InvalidCredentials` on both invalid username or password, and I actually find it better to not show which was invalid for security reasons.

Reply:
Yep, just add a generic error message, that's what I mean

Reply:
> Yep, just add a generic error message, that's what I mean
ah so something like "invalid email or password", got it.

>  actually, shouldn't do that bit of code I suggested, but rather use setError with no field somehow and use that as the error message on top of the form instead, because backend errors aren't even specific

form-level errors are removed directly after typing anything due to client side validation, see below comment